[All hosts](security) Enable automatic saving of passwords in webview2 (#5080)
OfficeDev/office-js-docs-pr: docs/develop/auth-save-passwords.md

---
title: Automatically save passwords in Microsoft Edge WebView2
description: Learn how to enable WebView2 to save passwords when users sign in using your add-in.
ms.localizationpriority: medium
ms.date: 03/10/2025
---

# Enable automatic password saving in Microsoft Edge WebView2

> If you use a password form to sign in the user, most browsers can automatically save passwords on behalf of the user when they sign in. This helps users avoid memorizing complex passwords with repeated sign-in prompts. When your add-in is loaded in Microsoft Office on Windows, Microsoft Edge Webview2 hosts your add-in. WebView2 also supports automatic password saving. To enable automatic password saving, add HTML input controls for the user name ans password similar to the following HTML.

Reviewer: A few suggestions:
- I think the "if" part of the first sentence is unnecessary. That users are inputting their credentials in an explicit form shouldn't matter.
- Reworded the second sentence, partially because repeatedly entering your password would help you memorize it.
- Flipped the third and fourth sentences to help with the reading flow. This way, it goes from abstract to add-in specific.
- The style guide indicates username is one word in this context.
```suggestion
Most browsers can automatically save passwords on behalf of the user when they sign in. This helps users manage passwords in a secure environment. Microsoft Edge WebView2 also supports automatic password saving. When your add-in is loaded in Microsoft Office on Windows, Webview2 hosts your add-in. To enable automatic password saving, add HTML input controls for the username and password, as shown in the following HTML.
```

> If you use a password form to sign in the user, most browsers can automatically save passwords on behalf of the user when they sign in. This helps users avoid memorizing complex passwords with repeated sign-in prompts. When your add-in is loaded in Microsoft Office on Windows, Microsoft Edge Webview2 hosts your add-in. WebView2 also supports automatic password saving. To enable automatic password saving, add HTML input controls for the user name ans password similar to the following HTML.

Reviewer: We explicitly say "Windows", but this applies to Edge web clients too, right?

Author: Information here only applies to webview2.

```html
<div>
    <label for="username">Username:</label><br/>
    <input type="text" id="username" name="username" /><br/>
    
    <label for="password">Password:</label><br/>
    <input type="password" id="password" name="password" /><br/>
    
    <button id="btn" type="button">Sign in</button>
</div>
```

> In the button click event handler for the sign-in button, call the authentication library of your choice to sign in the user. Once the sign-in is complete, redirect to a new web page. When WebView2 see the redirect, and the user name and password, it prompts the user to offer to automatically save the user name and password. The following code shows how to handle the sign-in button click event.

Reviewer: A couple suggestions for typos, grammar, and avoiding repeated words.
```suggestion
In the button click event handler for the sign-in button, call the authentication library of your choice to sign in the user. Once the sign-in is complete, redirect to a new web page. When WebView2 sees the redirect, and the username and password, it prompts the user to offer to automatically save the credentials. The following code shows how to handle the sign-in button click event.
```

```javascript
async function btnSignIn() {

>   // Get the user name and password credentials entered by the user.

Reviewer: ```suggestion
  // Get the username and password credentials entered by the user.
```

>   const username = document.getElementById("username").value

Reviewer: ```suggestion
  const username = document.getElementById("username").value;
```

>   const pwd = document.getElementById("password").value

Reviewer: ```suggestion
  const pwd = document.getElementById("password").value;
```

  try {

>     //Sign in the user. This is a placeholder for the actual sign-in logic.

Reviewer: ```suggestion
    // Sign in the user. This is a placeholder for the actual sign-in logic.
```

    await signInUser(username, pwd);

    // Redirect to a success page to trigger the password autosave.
    window.location.href = "/home.html";
  }
  catch (error) {
    console.error("Sign in failed: " + error);
    return;
  }
}
```

## How the user manages passwords

> When the user enters a new password in your add-in, and your add-in redirects to a new web page, WebView2 asks the user if they want their username and password saved. The next time your add-in prompts for credentials, WebView2 automatically enters the user's account info.

Reviewer: This sounds more natural to me, but I'm not sure why.
```suggestion
When the user enters a new password in your add-in, and your add-in redirects to a new web page, WebView2 asks the user if they want to save their username and password. The next time your add-in prompts for credentials, WebView2 automatically enters the user's account information.
```

> When the user enters a new password in your add-in, and your add-in redirects to a new web page, WebView2 asks the user if they want their username and password saved. The next time your add-in prompts for credentials, WebView2 will finish filling in the user's account info.

Reviewer: Suggestion to avoid the gerund "filling". 

```suggestion
When the user enters a new password in your add-in, and your add-in redirects to a new web page, WebView2 asks the user if they want their username and password saved. The next time your add-in prompts for credentials, WebView2 automatically enters the user's account info.
```

> :::image type="content" source="../images/edge-webview2-automatic-save-passwords.png" alt-text="The dialog from WebView2 prompting the user if they want to save their user name and password.":::

Reviewer: ```suggestion
:::image type="content" source="../images/edge-webview2-automatic-save-passwords.png" alt-text="The dialog from WebView2 prompting the user if they want to save their username and password.":::
```

> If the user wants to remove saved passwords, they can do so by deleting The WebView2 local cache folder at `%LOCALAPPDATA%\Microsoft\Office\16.0\Wef\webview2\`. If your add-in relies on automatically saving passwords, you should document this folder location so users can remove their passwords.

Reviewer: Consider this simplification. 
```suggestion
Users remove saved passwords by deleting The WebView2 local cache folder at `%LOCALAPPDATA%\Microsoft\Office\16.0\Wef\webview2\`. If your add-in relies on automatically saving passwords, you should document this folder location so users can remove their passwords.
```

> If the user wants to remove saved passwords, they can do so by deleting The WebView2 local cache folder at `%LOCALAPPDATA%\Microsoft\Office\16.0\Wef\webview2\`. If your add-in relies on automatically saving passwords, you should document this folder location so users can remove their passwords.

Reviewer: If the cache is being [automatically cleared](https://learn.microsoft.com/en-us/office/dev/add-ins/testing/clear-cache#automatically-clear-the-cache), will the passwords be removed too?

Author: Yes.
